File: service/shared/models/message.go
package models

import (
	"database/sql"
	"encoding/base64"
	"encoding/json"
	"time"
	"wasa/service/shared/helper"
)

type Message struct {
	ID             int           `json:"-"`
	Content        []byte        `json:"-"`
	ContentType    string        `json:"content_type"`
	SentTime       time.Time     `json:"sent_time"`
	EditedTime     sql.NullTime  `json:"-"`
	DeletedTime    sql.NullTime  `json:"-"`
	senderID       int           `json:"-"`
	ConversationID int           `json:"conversation_id"`
	RepliedTo      sql.NullInt64 `json:"-"`
	IsForwarded  bool `json:"-"`
	Sender 	   User          `json:"user"`
	RepliedToMessage *Message    `json:"-"`
    Reactions      []Reaction    `json:"-"`
}

func (m Message) MarshalJSON() ([]byte, error) {

    type messageAlias struct {
        ID              int        `json:"message_id"`
        Content         string     `json:"content"`
        ContentType     string     `json:"content_type"`
        SentTime        time.Time  `json:"sent_time"`
        EditedTime      *time.Time `json:"edited_time,omitempty"`
        DeletedTime     *time.Time `json:"deleted_time,omitempty"`
        SenderID        int        `json:"-"`
        ConversationID  int        `json:"conversation_id"`
        RepliedTo       *int64     `json:"-"`
        IsForwarded     bool       `json:"is_forwarded,omitempty"`
        Sender          User       `json:"sender"`
        RepliedToMessage *struct {  
            ID          int    `json:"message_id"`
            Content     string `json:"content"`
            ContentType string `json:"content_type"`
        } `json:"replied_to_message,omitempty"`
        Reactions       []Reaction `json:"reactions,omitempty"`
    }

    var content string
    if m.ContentType == "image" {
        content = base64.StdEncoding.EncodeToString(m.Content)
    } else {
        content = string(m.Content)
    }

    return json.Marshal(&messageAlias{
        ID:             m.ID,
        Content:        content,
        ContentType:    m.ContentType,
        SentTime:       m.SentTime,
        EditedTime:     helper.NullTimeToPtr(m.EditedTime),
        DeletedTime:    helper.NullTimeToPtr(m.DeletedTime),
        RepliedTo:      helper.NullInt64ToPtr(m.RepliedTo),
        IsForwarded:    m.IsForwarded,
        SenderID:       m.senderID,
        ConversationID: m.ConversationID,
        Sender:         m.Sender,
        Reactions:      m.Reactions,
        RepliedToMessage: func() *struct {
            ID          int    `json:"message_id"`
            Content     string `json:"content"`
            ContentType string `json:"content_type"`
        } {
            if m.RepliedToMessage == nil {
                return nil
            }
            
            var repliedContent string
            if m.RepliedToMessage.ContentType == "image" {
                repliedContent = base64.StdEncoding.EncodeToString(m.RepliedToMessage.Content)
            } else {
                repliedContent = string(m.RepliedToMessage.Content)
            }
            
            return &struct {
                ID          int    `json:"message_id"`
                Content     string `json:"content"`
                ContentType string `json:"content_type"`
            }{
                ID:          m.RepliedToMessage.ID,
                Content:     repliedContent,
                ContentType: m.RepliedToMessage.ContentType,
            }
        }(),
    })
}

File: webui/src/modules/message/models/message.ts
import { User, type UserDTO } from '@/modules/auth/models/user'
import type { Conversation } from '@/modules/conversation/models/conversation'
import type { Reaction } from './reaction'

export class Message {
  public messageId: number
  public content: Blob
  public contentType: string
  public sentTime: Date
  public editedTime?: Date
  public deletedTime?: Date
  public conversationId: number
  public repliedTo?: number
  public forwardedFrom?: number
  public reactions?: Reaction[]
  public sender: User

  constructor(data: MessageDTO) {
    this.messageId = data.message_id
    this.content = data.content
    this.contentType = data.content_type
    this.sentTime = new Date(data.sent_time)
    this.editedTime = data.edited_time ? new Date(data.edited_time) : undefined
    this.deletedTime = data.deleted_time ? new Date(data.deleted_time) : undefined
    this.conversationId = data.conversation_id
    this.repliedTo = data.replied_to
    this.forwardedFrom = data.forwarded_from
    this.sender = new User(data.sender)
  }

  static fromJSON(json: MessageDTO): Message {
    return new Message(json)
  }

  toJSON(): MessageDTO {
    return {
      message_id: this.messageId,
      content: this.content,
      content_type: this.contentType,
      sent_time: this.sentTime.toISOString(),
      edited_time: this.editedTime?.toISOString(),
      deleted_time: this.deletedTime?.toISOString(),
      conversation_id: this.conversationId,
      replied_to: this.repliedTo,
      forwarded_from: this.forwardedFrom,
      created_at: this.sentTime.toISOString(),
      sender: this.sender.toJSON(),
    }
  }

  isEdited(): boolean {
    return !!this.editedTime
  }

  isDeleted(): boolean {
    return !!this.deletedTime
  }
}

export interface MessageDTO {
  message_id: number
  content: Blob
  content_type: string
  sent_time: string
  edited_time?: string
  deleted_time?: string
  conversation_id: number
  replied_to?: number
  forwarded_from?: number
  created_at: string
  sender: UserDTO
}
